src/bit-fields-macros/src/utils.rs

```diff
@@ -0,0 +1,228 @@
+            "u16" => Ok(Self::U16),
+            "u32" => Ok(Self::U32),
+            "u64" => Ok(Self::U64),
+            "u128" => Ok(Self::U128),
+            _ => Err(DataTypeTryFromError),
+        }
+    }
+}
+
+/// Error type for `impl TryFrom<&str> for DataType`
+#[derive(Debug, PartialEq, Eq)]
+pub struct DataTypeTryFromError;
+
+impl fmt::Display for DataTypeTryFromError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "Bad backing data type given. Only accepted types are `u8`, `u16`, `u32`, `u64` and \
+             `u128`."
+        )
+    }
+}
+
+impl std::error::Error for DataTypeTryFromError {}
+
+/// Utility used to assembly the display string
+#[derive(Debug, Default)]
+pub struct MultiLineString(pub Vec<String>);
+
+impl MultiLineString {
+    /// Create new empty string.
+    pub fn new() -> Self {
+        Self(Vec::new())
+    }
+    /// Pushes `MultiLineString` on to `self`, if the given `other` is deeper than `self` it will
+    /// expand down to accommodate.
+    // None of these indexes or arithmetic operations will ever panic, thus we can safely use
+    // unchecked operations here. Since unchecked arithmetic is currently unstable we just allow
+    // integer arithmetic for now.
+    #[allow(clippy::integer_arithmetic, clippy::arithmetic)]
+    pub fn push(&mut self, other: &Self) {
+        // SAFETY:
+        // Always safe.
+        unsafe {
+            let mut i = 0;
+            while i < std::cmp::min(self.0.len(), other.0.len()) {
+                self.0
+                    .get_unchecked_mut(i)
+                    .push_str(other.0.get_unchecked(i));
+                i += 1;
+            }
+            while i < other.0.len() {
+                self.0.push(other.0.get_unchecked(i).clone());
+                i += 1;
+            }
+        }
+    }
+    /// Pushes a static string onto `self`.
+    pub fn push_str(&mut self, s: &str) {
+        self.push(&MultiLineString::from(s));
+    }
+}
+
+impl From<&str> for MultiLineString {
+    fn from(s: &str) -> Self {
+        Self(s.split('\n').map(String::from).collect())
+    }
+}
+
+impl From<MultiLineString> for String {
+    fn from(mls: MultiLineString) -> Self {
+        let mut iter = mls.0.into_iter();
+        if let Some(mut base) = iter.next() {
+            for next in iter {
+                base.push('\n');
+                base.push_str(&next);
+            }
+            base
+        } else {
+            String::new()
+        }
+    }
+}
+
+impl fmt::Display for MultiLineString {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        for line in &self.0 {
+            writeln!(f, "{}", line)?;
+        }
+        Ok(())
+    }
+}
+
+/// An iterator over sub-slices of a given slice e.g. `[0..]`, `[1..]`, .., `[slice.len()-1..]`.
+///
+/// The last slice will always contain 1 element.
+#[derive(Debug, Clone, Copy)]
+pub struct SliceIter<'a, T> {
```

Comment:
How is this different to std::slice::Iter?

Reply:
I was not aware of `std::slice::Iter` when I wrote this, so it may be a worthwhile replacement here.

To give 2 examples:
`SliceIter`:
```rust
let arr = [1,2,3];
let mut iter = SliceIter::from(&arr);
assert!(iter.next(),Some(&[1,2,3]));
assert!(iter.next(),Some(&[2,3]));
assert!(iter.next(),Some(&[3]));
assert!(iter.next(),None);
```
`std::slice:Iter`:
```rust
let arr = [1,2,3];
let mut iter = SliceIter::from(&arr);
assert!(iter.as_slice(),&[1,2,3]);
assert!(iter.next(),Some(1));
assert!(iter.as_slice(),&[2,3]);
assert!(iter.next(),Some(2));
assert!(iter.as_slice(),&[3]);
assert!(iter.next(),Some(3));
assert!(iter.as_slice(),&[]);
assert!(iter.next(),None);
```
So `SliceIter` offers a moderately nicer interface.